object-store/src/main/java/org/hypertrace/config/objectstore/ExternalContextIdentifiedObjectStore.java

```diff
@@ -0,0 +1,157 @@
+import org.hypertrace.config.service.v1.GetConfigRequest;
+import org.hypertrace.config.service.v1.UpsertConfigRequest;
+import org.hypertrace.config.service.v1.UpsertConfigResponse;
+import org.hypertrace.core.grpcutils.context.RequestContext;
+
+/**
+ * ExternalContextIdentifiedObjectStore is an abstraction over the grpc api for config
+ * implementations working with single object with identifier obtained from outside the data object
+ * - a pattern often used for a tenant-wide config.
+ *
+ * @param <T>
+ */
+@Slf4j
+public abstract class ExternalContextIdentifiedObjectStore<T> {
+  private final ConfigServiceBlockingStub configServiceBlockingStub;
+  private final String resourceNamespace;
+  private final String resourceName;
+  private final Optional<ConfigChangeEventGenerator> configChangeEventGeneratorOptional;
+
+  protected ExternalContextIdentifiedObjectStore(
+      ConfigServiceBlockingStub configServiceBlockingStub,
+      String resourceNamespace,
+      String resourceName,
+      ConfigChangeEventGenerator configChangeEventGenerator) {
+    this.configServiceBlockingStub = configServiceBlockingStub;
+    this.resourceNamespace = resourceNamespace;
+    this.resourceName = resourceName;
+    this.configChangeEventGeneratorOptional = Optional.of(configChangeEventGenerator);
+  }
+
+  protected ExternalContextIdentifiedObjectStore(
+      ConfigServiceBlockingStub configServiceBlockingStub,
+      String resourceNamespace,
+      String resourceName) {
+    this.configServiceBlockingStub = configServiceBlockingStub;
+    this.resourceNamespace = resourceNamespace;
+    this.resourceName = resourceName;
+    this.configChangeEventGeneratorOptional = Optional.empty();
+  }
+
+  protected abstract Optional<T> buildDataFromValue(Value value);
+
+  protected abstract Value buildValueFromData(T data);
+
+  protected Value buildValueForChangeEvent(T data) {
+    return this.buildValueFromData(data);
+  }
+
+  protected String buildClassNameForChangeEvent(T data) {
+    return data.getClass().getName();
+  }
+
+  public Optional<T> getData(RequestContext context) {
+    try {
+      Value value =
+          context.call(
+              () ->
+                  this.configServiceBlockingStub
+                      .getConfig(
+                          GetConfigRequest.newBuilder()
+                              .setResourceName(this.resourceName)
+                              .setResourceNamespace(this.resourceNamespace)
+                              .build())
+                      .getConfig());
+      T data = this.buildDataFromValue(value).orElseThrow(Status.INTERNAL::asRuntimeException);
+      return Optional.of(data);
+    } catch (Exception exception) {
+      if (Status.fromThrowable(exception).equals(Status.NOT_FOUND)) {
+        return Optional.empty();
+      }
+      throw exception;
+    }
+  }
+
+  public ConfigObject<T> upsertObject(RequestContext context, String dataContext, T data) {
```

Comment:
Be careful with naming. What we call dataContext on upsert is called id on delete.